### ISSUE TITLE
wal-e unexpected behavior when encountering PG deleted files

### ISSUE THREAD

**Issue by LeoCavaille** [wal-e/wal-e#228]
Recently on one of our Postgres instances we deleted a lot of schemas, hence freeing a lot of disk space.
The daily wal-e `backup-push` behaved very weirdly, it was still running after more than 1.5 days when usually it would take around 6h. When looking more precisely we could actually see that the deletes started just before kicking off the wal-e backup and finished 20min later:

![screen shot 2016-03-17 at 3 05 27 pm](https://cloud.githubusercontent.com/assets/1788830/13848377/c87550d8-ec51-11e5-8c53-0b58549dc31b.png)

The wal-e backup process looked very active still more than 1.5 days later, so when stracing what files it was currently working on, you could see that definitely something was wrong:

```
└─$  sudo strace -p 103750 -e open
Process 103750 attached with 2 threads
[pid 103750] open("/tmp/tmphDg_V3", O_RDWR|O_CREAT|O_EXCL|O_NOFOLLOW, 0600) = 4
[pid 103750] open("/mnt/postgres/9.3/datadog/base/16384/305961.26", O_RDONLY) = 7
[pid 103750] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=115219, si_status=0, si_utime=116, si_stime=6} ---
[pid 103750] open("/usr/local/lib/python2.7/dist-packages/boto/cacerts/cacerts.txt", O_RDONLY) = 8
[pid 103750] open("/tmp/tmp2Q_l_j", O_RDWR|O_CREAT|O_EXCL|O_NOFOLLOW, 0600) = 4
[pid 103750] open("/mnt/postgres/9.3/datadog/base/16384/305961.26", O_RDONLY) = 7
[pid 103750] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=115303, si_status=0, si_utime=110, si_stime=7} ---
[pid 103750] open("/usr/local/lib/python2.7/dist-packages/boto/cacerts/cacerts.txt", O_RDONLY) = 8
[pid 103750] open("/tmp/tmpIPezjs", O_RDWR|O_CREAT|O_EXCL|O_NOFOLLOW, 0600) = 4
[pid 103750] open("/mnt/postgres/9.3/datadog/base/16384/305961.26", O_RDONLY) = 7
[pid 103750] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=115377, si_status=0, si_utime=122, si_stime=6} ---
[pid 103750] open("/usr/local/lib/python2.7/dist-packages/boto/cacerts/cacerts.txt", O_RDONLY) = 8
[pid 103750] open("/tmp/tmpGaupyv", O_RDWR|O_CREAT|O_EXCL|O_NOFOLLOW, 0600) = 4
[pid 103750] open("/mnt/postgres/9.3/datadog/base/16384/305961.26", O_RDONLY) = 7
[pid 103750] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=115446, si_status=0, si_utime=128, si_stime=7} ---

[repeats ...]
```

You can see that it's opening always the same file. When analyzing the code, especially the piece [where it creates tar partitions](https://github.com/wal-e/wal-e/blob/master/wal_e/tar_partition.py#L372-L389) before uploading. The highlighted part show the exception caught whenever it processes a file that has been unlinked.
In our case, lots of file were unlinked as several thousand schemas were dropped from PG (totaling around 700 GB) between the moment wal-e created the file list and inspecting files to assemble tar partitions.

According to the comment it's OK to just skip them because they can be replayed via the WAL (sounds right), but in practice it's not skipped due to the current code, it keeps on moving to adding this "file" to the partition. This is because in case the `OSError` exception happens `et_info` holds the value of the last loop iteration.
I believe this leads to the behavior witnessed on strace which seems undesirable.

I've created [a branch](https://github.com/wal-e/wal-e/compare/master...LeoCavaille:deleted_files?expand=1) with a fix (sic, details are in the commit message way longer than the commit itself...), but lacking general knowledge of the code, I don't know what edge cases or unexpected behavior this could provoke.

An obvious workaround is to kill the current backup and restart a new backup after the deletion is complete.

Thanks for your help,

**Comment by fdr** [comment]
That is a bug! Can you send a PR?

**Comment by fdr** [comment]
This was fixed in in #236
